### ISSUE TITLE
Make SMC more efficient by allocating memory at the model-layer

### ISSUE THREAD

**Issue by dereklhansen** [dereklhansen/SMC.jl#19]
One area for improvement is to not allocate memory when new particles are sampled. We currently don't do this to keep compatibility with autodiff.

The way to do this is actually at the model layer (e.g. LinearGaussian) rather than at the SMCModel layer

```
function rp_base(m::LinearGaussian, rng, Zs, t)
    μ, Σ = calc_prop(m, Zs, t)
    @assert isposdef(Σ)
    Zs_new = μ + rand(rng, MvNormal(Σ), size(Zs, 2))
    return Zs_new
end
```

turns into 

```
function rp_base(m::LinearGaussian, rng, Zs, t)
    μ, Σ = calc_prop(m, Zs, t)
    @assert isposdef(Σ)
    eps = rand!(rng, MvNormal(Σ), m._eps_old)
    eps .+= μ
    m._eps_old = m._eps_new # swich the underlying storage
    m._eps_new = eps
    return Zs_new
end
```

**Comment by dereklhansen** [comment]
Maybe implement a proof-of-concept in a new model (LinearGaussianCached ?) to compare performance

**Comment by dereklhansen** [comment]
Closed by #20
